### PR TITLE
Refactor of best_season and worst_season methods in TeamStatistics module

### PR DESCRIPTION
**Pull request by c-bartell**
Refactored methods to extract redundant code

- Refactored separate_seasons_by_team_id to use reduce
- Extracted results_by_season and season_totals from best_season and worst_season
    - Added unit tests for each
- Removed some notes and un needed whitespace
- Made some edits suggested by Rubocop
- Module test suite and spec harness passing

### CODE REVIEW

File: lib/team_statistics.rb [hunk 1]
```diff
@@ -20,37 +20,28 @@ def games_by_team_id(team_id, games_array = games)
...
 
   def separate_games_by_season_id(games_array = games)
-    season_hash = {}
-    games_array.each do |game|
-      if season_hash[game.season] == nil
-        season_hash[game.season] = [game]
-      else
-        season_hash[game.season] << game
-      end
+    games_array.reduce({}) do |seasons, game|
```
Comment (cpfergus1) [inline lib/team_statistics.rb:27]: I really like how you refactored here and moved out reused code

File: lib/team_statistics.rb [hunk 2]
```diff
@@ -63,124 +54,107 @@ def game_stats_by_team_id(team_id)
...
+  def best_season(team_id)
+    totals = season_totals(results_by_season(team_id))
+    totals.keys.max_by do |season|
+      totals[season][:average]
     end
+  end
 
-    totals_by_season.keys.min_by do |season|
-      totals_by_season[season][:average]
+  def worst_season(team_id)
```
Comment (cpfergus1) [inline lib/team_statistics.rb:104]: much more concise and easier to read!

**Review by cpfergus1 — APPROVED** [review]
Really good refactor, I like it!

File: lib/team_statistics.rb [hunk 3]
```diff
@@ -20,37 +20,28 @@ def games_by_team_id(team_id, games_array = games)
...
   end
 
   def separate_games_by_season_id(games_array = games)
-    season_hash = {}
-    games_array.each do |game|
-      if season_hash[game.season] == nil
-        season_hash[game.season] = [game]
-      else
-        season_hash[game.season] << game
-      end
```
Comment (s-steel) [inline lib/team_statistics.rb:35]: Nice use of reduce!

File: lib/team_statistics.rb [hunk 4]
```diff
@@ -63,124 +54,107 @@ def game_stats_by_team_id(team_id)
...
-      game.goals
-    end.goals
+    game_stats_by_team_id(team_id).max_by(&:goals).goals
   end
 
   def fewest_goals_scored(team_id)
-    game_stats_by_team_id(team_id).min_by do |game|
-      game.goals
-    end.goals
+    game_stats_by_team_id(team_id).min_by(&:goals).goals
```
Comment (s-steel) [inline lib/team_statistics.rb:61]: Great refactoring.  Nice and short

File: lib/team_statistics.rb [hunk 5]
```diff
@@ -63,124 +54,107 @@ def game_stats_by_team_id(team_id)
...
-      end
-    end
-    totals_by_season = {}
-    results_by_season.each do |season, games|
-      totals_by_season[season] = {}
-      totals_by_season[season][:total] = games.length
-      totals_by_season[season][:wins] = games.select do |game|
-        game.result == "WIN"
-      end.length
-      totals_by_season[season][:average] = (totals_by_season[season][:wins].to_f / totals_by_season[season][:total].to_f).round(2)
```
Comment (s-steel) [inline lib/team_statistics.rb:125]: This looks a lot better.

**Review by s-steel — APPROVED** [review]
Looks great.  Nice refactoring.

**Pull request merged**